### PR TITLE
ST_EXTENT aggregation (#104659)

### LINKED ISSUE #104659: Support ST_EXTENT_AGG in ESQL
### Description

The simplest spatial aggregation once both points and shapes are supported is `ST_EXTENT`. This is also the only aggregation being considered for non-licensing due to its potential usage in Kibana. The function is similar to [`ST_ENVELOPE`](https://github.com/elastic/elasticsearch/issues/104875), but instead of generating the bounds for a geometry argument, it aggregates over all geometries for that field, and produces a combined bounding box.

For reference, the PostGIS documentation is https://postgis.net/docs/ST_Extent.html

### PR DESCRIPTION
**Pull request by GalLalouche**
Support `ST_EXTENT` aggregation (Issue #104659)

This PR adds support for `ST_EXTENT` aggregation, i.e., computing a bounding box over a set of points/shapes (Cartesian or geo). (Note the difference between this aggregation and the already implemented *scalar *function `ST_EXTENT`). 

* This isn't a very efficient implementation, and future PRs will attempt to read these extents directly from the doc values.
* We currently always use longitude wrapping, i.e., we may wrap around the dateline for a smaller bounding box. Future PRs will let the user control this behavior.

Fixes #104659

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/EsqlFunctionRegistry.java [hunk 1]
```diff
@@ -346,6 +347,7 @@ private static FunctionDefinition[][] functions() {
             // spatial
             new FunctionDefinition[] {
                 def(SpatialCentroid.class, SpatialCentroid::new, "st_centroid_agg"),
+                def(SpatialStExtent.class, SpatialStExtent::new, "st_extent"),
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/EsqlFunctionRegistry.java:350]: As much as possible, these should be listed alphabetically.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java [hunk 2]
```diff
@@ -22,26 +25,34 @@
  * select the best one.
  */
 public abstract class SpatialAggregateFunction extends AggregateFunction {
-    protected final boolean useDocValues;
+    protected final FieldExtractPreference preference;
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java:28]: Can you find a different name for `preference`? Something that's more explanatory.

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/60_usage.yml [hunk 3]
```diff
@@ -25,142 +25,142 @@ setup:
...
-  - set: {esql.features.keep: keep_counter}
-  - set: {esql.features.limit: limit_counter}
-  - set: {esql.features.mv_expand: mv_expand_counter}
-  - set: {esql.features.rename: rename_counter}
-  - set: {esql.features.row: row_counter}
-  - set: {esql.features.show: show_counter}
-  - set: {esql.features.sort: sort_counter}
-  - set: {esql.features.stats: stats_counter}
-  - set: {esql.features.where: where_counter}
+  - set: { esql.features.dissect: dissect_counter }
```
Comment (astefan) [inline x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/60_usage.yml:112]: There are so many unnecessary changes in this file related to whitespacing, unrelated to the actual meat of the PR. Would you mind keeping just the relevant bits?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java [hunk 4]
```diff
@@ -22,26 +25,34 @@
...
 
-    protected SpatialAggregateFunction(Source source, Expression field, Expression filter, boolean useDocValues) {
+    protected SpatialAggregateFunction(Source source, Expression field, Expression filter, FieldExtractPreference values) {
         super(source, field, filter, emptyList());
-        this.useDocValues = useDocValues;
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java:29]: Given this change here, I'm wondering if there are no serialization issues regarding BWC.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/AggregateFunction.java [hunk 5]
```diff
@@ -29,7 +29,6 @@
  * A type of {@code Function} that takes multiple values and extracts a single value out of them. For example, {@code AVG()}.
  */
 public abstract class AggregateFunction extends Function {
-
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/AggregateFunction.java:32]: Unnecessary change.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java [hunk 6]
```diff
@@ -22,26 +25,34 @@
...
         super(in);
         // The useDocValues field is only used on data nodes local planning, and therefor never serialized
-        this.useDocValues = useDocValues;
+        this.preference = preference;
     }
 
     public abstract SpatialAggregateFunction withDocValues();
 
+    @Override
+    public boolean checkLicense(XPackLicenseState state) {
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java:44]: Are we testing this anywhere?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java (hunk 2 above)
Comment (GalLalouche) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java:28]: As in, `fieldExtractPreference`?

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/60_usage.yml (hunk 3 above)
Comment (GalLalouche) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/60_usage.yml:112]: Sorry, I think it's the automatic formatter. I'll try to revert it.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java (hunk 4 above)
Comment (GalLalouche) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java:29]: Could you please elaborate?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/EsqlFunctionRegistry.java (hunk 1 above)
Comment (GalLalouche) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/EsqlFunctionRegistry.java]: Done.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/AggregateFunction.java (hunk 5 above)
Comment (GalLalouche) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/AggregateFunction.java]: Sorry, force of habit :)

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java (hunk 6 above)
Comment (GalLalouche) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java:44]: Could you please direct me to an example license test?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java (hunk 4 above)
Comment (astefan) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java:29]: Some things are serialized in communication node to node. I am wondering if the change from boolean to FieldExtractPreference has any connection with serialization and, implicitly, BWC (backwards compatibility). Have you looked into that?

Comment (GalLalouche) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java:29]: Given the comment in the file:
```
        // The useDocValues field is only used on data nodes local planning, and therefor never serialized
```
Wouldn't this mean the isn't any issue regarding serialization?

Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java:29]: Correct. This field is only every planned or modified on the data nodes, and is therefor never serialized.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java (hunk 6 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialAggregateFunction.java:44]: `CheckLicenseTests` tests the licensing checks in general. I wonder if there is a way to write a negative test for our code? We could look in the SpatialPlugin for similar tests.

File: libs/geo/src/test/java/org/elasticsearch/geometry/utils/SpatialEnvelopeVisitorTests.java [hunk 7]
```diff
@@ -48,7 +49,7 @@ public void testVisitGeoShapeNoWrap() {
     public void testVisitGeoShapeWrap() {
         for (int i = 0; i < 1000; i++) {
             var geometry = GeometryTestUtils.randomGeometryWithoutCircle(0, true);
-            var bbox = SpatialEnvelopeVisitor.visitGeo(geometry, false);
+            var bbox = SpatialEnvelopeVisitor.visitGeo(geometry, WrapLongitude.NO_WRAP);
```
Comment (craigtaverner) [inline libs/geo/src/test/java/org/elasticsearch/geometry/utils/SpatialEnvelopeVisitorTests.java:52]: I think this should be wrapping, according to the test method name.

Comment (GalLalouche) [reply libs/geo/src/test/java/org/elasticsearch/geometry/utils/SpatialEnvelopeVisitorTests.java:52]: Then the test itself needs to change because it's failing when it's set to `WRAP`

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/RectangleMatcher.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+package org.elasticsearch.xpack.esql.expression;
+
+import org.elasticsearch.compute.aggregation.spatial.PointType;
+import org.elasticsearch.geometry.Rectangle;
+import org.hamcrest.Description;
+import org.hamcrest.Matchers;
+import org.hamcrest.TypeSafeMatcher;
+
+public class RectangleMatcher extends TypeSafeMatcher<Rectangle> {
```
Comment (craigtaverner) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/RectangleMatcher.java:16]: Add javadoc

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/WellKnownBinaryBytesRefMatcher.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.geometry.Geometry;
+import org.elasticsearch.geometry.utils.GeometryValidator;
+import org.elasticsearch.geometry.utils.WellKnownBinary;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.TypeSafeMatcher;
+
+public class WellKnownBinaryBytesRefMatcher<G extends Geometry> extends TypeSafeMatcher<BytesRef> {
```
Comment (craigtaverner) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/WellKnownBinaryBytesRefMatcher.java:18]: Add javadoc

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java [hunk 10]
```diff
@@ -2789,6 +2791,68 @@ public void testSpatialTypesAndStatsUseDocValues() {
...
+     *         \_FieldExtractExec[location{f}#48][location{f}#48]
+     *           \_EsQueryExec[airports], indexMode[standard], query[{"exists":{"field":"location","boost":1.0}}][
+     * _doc{f}#79], limit[], sort[] estimatedRowSize[25]
+     * </code>
+     * Note the FieldExtractExec has 'location' set for stats: FieldExtractExec[location{f}#9][location{f}#9]
+     * <p>
+     * Also note that the type converting function is removed when it does not actually convert the type,
+     * ensuring that ReferenceAttributes are not created for the same field, and the optimization can still work.
+     */
+    public void testSpatialTypesAndStatsExtentUseDocValues() {
```
Comment (craigtaverner) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:2822]: Perhaps a test that has both centroid and extent aggregations?

**Review by craigtaverner — CHANGES_REQUESTED** [review]

File: libs/geo/src/test/java/org/elasticsearch/geometry/utils/SpatialEnvelopeVisitorTests.java (hunk 7 above)
Comment (craigtaverner) [reply libs/geo/src/test/java/org/elasticsearch/geometry/utils/SpatialEnvelopeVisitorTests.java:52]: OK. Perhaps add a TODO comment.

**Review by craigtaverner — APPROVED** [review]
LGTM. I think some PhysicalPlanOptimizerTests over shapes would be good, but that can be done in the next PR where we do shape optimization.
